lib/IRGen/IRGenMangler.cpp

```diff
@@ -171,12 +178,21 @@ mangleSymbolNameForSymbolicMangling(const SymbolicMangling &mangling) {
 
   for (auto &symbol : mangling.SymbolicReferences) {
     // Fill in the placeholder space with something printable.
-    auto dc = symbol.first;
+    auto referent = symbol.first;
     auto offset = symbol.second;
-    Storage[prefixLen + offset] = Storage[prefixLen + offset+1] =
-      Storage[prefixLen + offset+2] = Storage[prefixLen + offset+3] = '_';
+    Storage[prefixLen + offset]
+      = Storage[prefixLen + offset+1]
+      = Storage[prefixLen + offset+2]
+      = Storage[prefixLen + offset+3]
+      = Storage[prefixLen + offset+4]
+      = '_';
     Buffer << ' ';
-    appendContext(dc);
+    if (auto ty = referent.dyn_cast<const NominalTypeDecl*>())
+      appendContext(ty);
+    else if (auto proto = referent.dyn_cast<const ProtocolDecl*>())
+      appendContext(proto);
```

Comment:
Given that a `ProtocolDecl` is-a `NominalTypeDecl`, we have to be excessively careful with that `referent` type. Do we need the protocol vs. nominal type distinction here?

Reply:
Probably not. I started out giving type and protocols different symbolic reference discriminators, but that ended up being unnecessary.